ArduCopter/GCS_Mavlink.cpp

```diff
@@ -829,14 +829,15 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_long_packet(const mavlink_command_
         // param3 : unused
         // param4 : unused
         if (packet.param2 > 0.0f) {
-            if (packet.param1 > 2.9f) { // 3 = speed down
-                copter.wp_nav->set_speed_down(packet.param2 * 100.0f);
-            } else if (packet.param1 > 1.9f) { // 2 = speed up
-                copter.wp_nav->set_speed_up(packet.param2 * 100.0f);
+            if ((packet.param1 > 2.9f) && (copter.flightmode->set_speed_down(packet.param2 * 100.0f))) { // 3 = speed down
+                return MAV_RESULT_ACCEPTED;
+            } else if ((packet.param1 > 1.9f) && (copter.flightmode->set_speed_up(packet.param2 * 100.0f))) { // 2 = speed up
+                return MAV_RESULT_ACCEPTED;
+            } else if (copter.flightmode->set_speed_xy(packet.param2 * 100.0f)) {
+                return MAV_RESULT_ACCEPTED;
             } else {
-                copter.wp_nav->set_speed_xy(packet.param2 * 100.0f);
+                return MAV_RESULT_FAILED;
```

Comment:
This needs to be:
```
         if (packet.param2 > 0.0f) {
            if (packet.param1 > 2.9f) {
              if (!copter.flightmode->set_speed_down(packet.param2 * 100.0f))) { // 3 = speed down
                return MAV_RESULT_FAILED;
              }
              return MAV_RESULT_ACCEPTED;
            }
            if ((packet.param1 > 1.9f) {
                if (!copter.flightmode->set_speed_up(packet.param2 * 100.0f))) { // 2 = speed up
                    return MAV_RESULT_FAILED;
                }
                return MAV_RESULT_ACCEPTED;
            }
            return MAV_RESULT_INVALID_PARAM2;
         }
         // param2 not set, so this is horizontal speed:
         if (!copter.flightmode->set_speed_xy(packet.param2 * 100.0f)) {
                return MAV_RESULT_FAILED;
         }
         return MAV_RESULT_ACCEPTED;
```

Your changes here changed the logic such that it would attempt up if the down clause failed, which isn't great :-)

Changes suggested here gives the mavlink client more information about what's going wrong.

Reply:
Thanks I think I have addressed this feedback with minimum changes.